Jack/Seed.fs

```diff
@@ -32,87 +58,69 @@ module Seed =
+        { s0 with Value = s0.Value + s0.Gamma }
     
-    /// Create a new 'Seed' using 'System.Random' to seed the generator.
+    /// Create a new random 'Seed'.
     let random () : Seed =
-        Random().Next() |> ofInt32
+        let s = System.DateTimeOffset.UtcNow.Ticks + 2L * goldenGamma
+        { Value = mix64 s
+          Gamma = mixGamma s + goldenGamma }
 
-    /// The possible range of values returned from 'next'.
-    let range : int * int =
-        1, 2147483562
-    
     /// Returns the next pseudo-random number in the sequence, and a new seed.
-    let next (Seed (s1, s2)) : int * Seed =
-        // TODO remove crashUnless
-        crashUnless (s1 >= 0) "s1 >= 0"
-        let k    = s1 / 53668
-        let s1'  = 40014 * (s1 - k * 53668) - k * 12211
-        let s1'' = if s1' < 0 then s1' + 2147483563 else s1'
-    
-        crashUnless (s2 >= 0) "s2 >= 0"
-        let k'   = s2 / 52774
-        let s2'  = 40692 * (s2 - k' * 52774) - k' * 3791
-        let s2'' = if s2' < 0 then s2' + 2147483399 else s2'
-    
-        let z    = s1'' - s2''
-        let z'   = if z < 1 then z + 2147483562 else z
-    
-        z', Seed (s1'', s2'')
-    
+    let next (s0 : Seed) : bigint * Seed =
+        let s = nextSeed s0
+        let n = mix64 s.Value
+        bigint n, nextSeed s
+
     /// Generate a random bigint in the specified range.
     let rec nextBigInt (lo : bigint) (hi : bigint) (seed : Seed) : bigint * Seed =
         if lo > hi then
             nextBigInt hi lo seed
         else
-            //
-            // Probabilities of the most likely and least likely result will differ
-            // at most by a factor of (1 +- 1/q). Assuming Seed is uniform, of
-            // course.
-            //
-            // On average, log q / log b more random values will be generated than
-            // the minimum.
-            //
-    
-            let genlo, genhi = range
-            let b = bigint genhi - bigint genlo + 1I
-    
-            let q = 1000I
-            let k = hi - lo + 1I
-            let magtgt = k * q
-    
-            // Generate random values until we exceed the target magnitude.
-            let rec loop mag v0 seed0 =
-                if mag >= magtgt then
-                    v0, seed0
-                else
-                    let x, seed1 = next seed0
-                    let v1 = v0 * b + (bigint x - bigint genlo)
-                    loop (mag * b) v1 seed1
-    
-            let v, seedN = loop 1I 0I seed
-
-            // TODO remove crashUnless
-            crashUnless (v >= 0I) "v >= 0I"
-            crashUnless (k >= 0I) "k >= 0I"
-            lo + v % k, seedN
+            let rec loop hilo (v0, seed0) =
+                if hilo > bigint.Zero then
+                    lo + v0 % hilo, seed0
+                else if v0 < lo || v0 >= hi then
+                    loop hilo (next seed0)
+                else v0, seed0
+            loop <| hi - lo + bigint.One <| next seed
```

Comment:
Can you explain how this works? I don't quite follow how it generates a number between the entire range `lo` to `hi`